ThirdpartyToolchain.cmake

```diff
@@ -0,0 +1,205 @@
+
+  # DOUBLE_CONVERSION INTERFACE_LINK_LIBRARIES REQUIRED
+  find_library(DOUBLE_CONVERSION_LIBRARY NAMES double-conversion)
+  if(DOUBLE_CONVERSION_LIBRARY)
+    list(APPEND FOLLY_LINK_LIBRARIES ${DOUBLE_CONVERSION_LIBRARY})
+  endif()
+
+  # LIBGFLAGS INTERFACE_LINK_LIBRARIES REQUIRED
+  find_library(LIBGFLAGS_LIBRARY NAMES gflags gflags_static)
+  if(LIBGFLAGS_LIBRARY)
+    list(APPEND FOLLY_LINK_LIBRARIES ${LIBGFLAGS_LIBRARY})
+  endif()
+
+  # OPENSSL INTERFACE_LINK_LIBRARIES REQUIRED
+  find_package(OpenSSL 1.1.1 MODULE REQUIRED)
+  if(OPENSSL_FOUND)
+    list(APPEND FOLLY_LINK_LIBRARIES ${OPENSSL_LIBRARIES})
+  endif()
+
+  # LIBEVENT INTERFACE_LINK_LIBRARIES REQUIRED
+  find_package(Libevent CONFIG)
+  if(LIBEVENT_FOUND)
+    list(APPEND FOLLY_LINK_LIBRARIES ${LIBEVENT_LIB})
+  endif()
+
+  # BOOST INTERFACE_LINK_LIBRARIES REQUIRED
+  find_package(Boost 1.51.0 MODULE
+  COMPONENTS
+    context
+    filesystem
+    program_options
+    regex
+    system
+    thread
+  REQUIRED
+  )
+  list(APPEND FOLLY_LINK_LIBRARIES ${Boost_LIBRARIES})
+
+  # ZSTD INTERFACE_LINK_LIBRARIES REQUIRED
+  find_library(ZSTD_LIBRARY_RELEASE NAMES zstd zstd_static)
+  if(ZSTD_LIBRARY_RELEASE)
+    list(APPEND FOLLY_LINK_LIBRARIES ${ZSTD_LIBRARY_RELEASE})
+  endif()
+
+  # LZ4 INTERFACE_LINK_LIBRARIES REQUIRED
+  find_library(LZ4_LIBRARY_RELEASE NAMES lz4)
+  if (LZ4_LIBRARY_RELEASE)
+    list(APPEND FOLLY_LINK_LIBRARIES ${LZ4_LIBRARY_RELEASE})
+  endif()
+
+  # LIBLZMA INTERFACE_LINK_LIBRARIES REQUIRED
+  find_package(LibLZMA MODULE)
+  set(FOLLY_HAVE_LIBLZMA ${LIBLZMA_FOUND})
+  if (LIBLZMA_FOUND)
+    list(APPEND FOLLY_LINK_LIBRARIES ${LIBLZMA_LIBRARIES})
+  endif()
+
+  #ZLIB INTERFACE_LINK_LIBRARIES REQUIRED
+  find_package(ZLIB MODULE)
+  if (ZLIB_FOUND)
+    list(APPEND FOLLY_LINK_LIBRARIES ${ZLIB_LIBRARIES})
+  endif()
+
+  add_dependencies(toolchain folly)
+  add_library(Folly::follybenchmark STATIC IMPORTED)
+  add_library(Folly::folly STATIC IMPORTED)
+
+  set_target_properties(Folly::follybenchmark
+                        PROPERTIES IMPORTED_LOCATION "${FOLLY_BENCHMARK_STATIC_LIB}"
+                        INTERFACE_INCLUDE_DIRECTORIES "${FOLLY_INCLUDE_DIR}"
+                        INTERFACE_LINK_LIBRARIES "${FOLLY_LINK_LIBRARIES}")
+  set_target_properties(Folly::folly
+                        PROPERTIES IMPORTED_LOCATION "${FOLLY_STATIC_LIB}"
+                        INTERFACE_INCLUDE_DIRECTORIES "${FOLLY_INCLUDE_DIR}"
+                        INTERFACE_LINK_LIBRARIES "${FOLLY_LINK_LIBRARIES}")
+
+  add_dependencies(Folly::folly folly)
+
+  set(FOLLY_LIBRARIES Folly::folly)
+endmacro()
+
+#===================================END FOLLY============================================
+
+macro(build_dependency DEPENDENCY_NAME)
+  if("${DEPENDENCY_NAME}" STREQUAL "folly")
+    build_folly()
+  else()
+    message(FATAL_ERROR "Unknown thirdparty dependency to build: ${DEPENDENCY_NAME}")
+  endif()
+endmacro()
+
+macro(resolve_dependency DEPENDENCY_NAME)
```

Comment:
nit: any reason why macro over function ? Also can we add some documentation here.

Reply:
The initial reason is that we use macros on ARROW. We can change it to functions, we would require some minor changes as some variables would be outside of the scope and they are required to be accessible, i.e:
```
set(FOLLY_LIBRARIES Folly::folly CACHE INTERNAL "Folly dependencies libraries")
```

Reply:
I have no strong opinions on macros vs functions here; I am curious why ARROW chose that , are there any benefits you folks saw ?

Reply:
I've found the original discussion when it was implemented about the question on why use macro over function: https://github.com/apache/arrow/pull/3688/files#r262737838
I don't think there was another benefit apart from helping exposing some variables globally due to the scope difference.